### DISCUSSION TITLE
Strange Error When Trying to Install HARMAN Adobe Air Runtimes on Mac OS X

### DISCUSSION THREAD

**Discussion by RobK88** [airsdk/Adobe-Runtime-Support#1890] [Q&A] [answered]
I have been trying to install the HARMAN Adobe Air Runtimes (version 33.1.1.743 and version 33.1.1.821) on my old Mac running Mac OS X Lion (10.7.5) without success.

The old Mac was running Adobe Air 32.0.0.125.  So I uninstalled it as recommended by HARMAN.

But when I try to install the HARMAN Adobe Air Runtimes (either version 33.1.1.743 or version 33.1.1.821) on my old Mac running Mac OS X Lion (10.7.5), the installer throws this strange error message:

```
Adobe AIR

This application requires a version of Adobe AIR which cannot be found.
Please download the latest version of the runtime from
http://www.adobe.com/go/getair,
or contact the application author for an updated version.
```

So I had to reinstall the old Adobe Air 32.0.0.125.

It looks like HARMAN Adobe Air Runtimes will not install on Mac OS X Lion (10.7.5).  But I was able to install the HARMAN Adobe Air runtimes on my other old Mac running Mountain Lion (10.8.5).

What are the system requirements for the HARMAN Adobe Air Runtimes?

Or is this just a bug in the HARMAN Adobe Air runtimes?

**Comment by ajwfrost** [comment, marked as answer]
Hi

Technically the builds should support 10.10 and later. We've actually found though that there are restrictions caused by the code signing, contrary to the Apple documentation this seems to prevent installation on 10.10 and 10.11... so basically, 10.12 and up would be best.

I'm not sure how it ended up running on 10.85! but it may be a quirk of some code-signing/gatekeeper functionality perhaps. There is also a possibility that the runtime will hit an API that's not available on the older macOS version and would then just exit, we don't actually have such an old machine to try things on now..

The only thing I can suggest perhaps is to try removing the code-signing and see if that makes a difference. (I believe it's all x86_64-based already...? and supports universal binaries, all that seems to be from 10.4...)  

thanks

**Comment by RobK88** [comment]
Thanks for the info.  If possible, I would add the system requirements to the HARMAN download pages.
I will try removing the code signing using the great free utility called "unsign" from steakknife.  See https://github.com/steakknife/unsign
But I doubt it is a gatekeeper issue.  It looks like the HARMAN Air Runtime Installer is looking for an API that is not available in Lion, so it throws this strange error instead.

FYI -- Yes, Mac OS X Tiger was actually the first Mac OS X release to run on Intel processors and it even add some support for 64-bit applications, so long as these did not utilize any of the GUI libraries (and obviously only on machines with 64-bit processors).  Real x86_64 support came in Leopard.  Tiger was also the first Mac OS X release to run PPC binaries on Intel hardware using Rosetta.  In other words, Lion ran x86 binaries, some x86_64 binaries, PPC binaries (and some universal binaries that contained 32 and 64 bit binaries and FAT binaries that contained 32-bit x86 Intel and PPC binaries).  Tiger was also the last Mac OS X version to support the Classic Environment – a Mac OS 9 compatibility layer – and PowerPC G3 processors.

As noted, full x86_64 Intel support came in Leopard and was improved in Snow Leopard.

Snow Leopard could only be installed on Macs with Intel processors (32-bit or 64-bit) and was the last Mac OS X release that ran both Intel and PPC binaries (via Rosetta).   (i.e. PPC binary support was dropped in Lion).    Snow Leopard also provided both 32-bit and 64-bit Intel kernels.

Lion could only be installed on Macs with 64-bit Intel processors and was the last release that maintained both 32-bit and 64-bit Intel kernels.  PPC binary support was dropped in Lion.

Mountain Lion deprecated the use of the Carbon API used for developing Mac OS X applications.  (Carbon lingered around and was eventually discontinued and removed in Catalina).

And Mojave was the last release that supported running both Intel x86 and x86_64 binaries.  (i.e. 32-bit Intel support was dropped in Catalina).

Catalina only runs x86_64 binaries and universal binaries that contained 32 and 64 bit Intel binaries.  (Only the 64-bit Intel binary in the universal binary will be used).   In Catalina, the Carbon API was also officially discontinued and removed, leaving Cocoa as the sole primary API for developing macOS applications.  (This makes sense since Carbon was only 32-bit anyway).

Big Sur was the first release to support both Intel x86_64 and ARM64 binaries and universal binaries 2 (that include both x86_64 and ARM64 binaries).  Intel x86_64 binaries can also run on Apple Silicon Macs under Big Sur via Rosetta 2.  (An option also exists to force a universal binary 2 to run as x86-64 code through Rosetta 2, even on an ARM-based machine).  The original universal binaries (that contain both 32 bit and 64 bit Intel binaries) will also run but only the 64-bit Intel binary would be used. 

P.P.S.  Apple's terminology can be confusing:

Fat binaries --- Intel x86 and PPC binaries.
Universal binaries -- Intel x86 and X86_64 binaries
Universal binaries -- Intel x86_64 and ARM64 binaries

So people call the last universal binary type -  universal binary 2!

**Reply by ajwfrost** [reply]
Thanks, that's a good bit of history! Will be interesting to see whether Apple end up completely dropping the Intel/AMD architecture support and focus just on their own ARM-based silicon...

I had expected that the binaries would just not run on an old macOS version, since there's a field in there which gives the minimum requirement. Seems odd to launch the app and then close it due to an API failure ... or perhaps this is a result of how the package is architected i.e. an initial executable which then loads in the AIR runtime as a framework library -> if the initial executable doesn't have the 'minimum OS' setting, then it could end up just failing at the "loading the runtime" stage....   and yet, if it's working on Mountain Lion, that implies that the setting is wrong (or this theory is wrong).... we can take a quick look at that!

thanks

**Comment by RobK88** [comment]
Well, I can confirm that the latest HARMAN Air Runtime (Version 33.1.1.821) does install on Mountain Lion.  But it does not work.  I tried installing some .air applications but each time the air installer would crash.  So I uninstalled it and reverted back to the last Adobe Air runtime released by Adobe (32.0.0.125).  

It looks like there is a bug in the latest HARMAN Air Runtime (Version 33.1.1.821).  It should not install on Mountain Lion (10.8.5).  And on Lion it throws the strange error noted earlier when trying to install the Runtime.  In both cases, a message really should be displayed stating that HARMAN Air Runtime requires 10.10 and later.

P.S.  Given Apple's history and business model, it is extremely likely that Apple will eventuality drop support for x86_64 binaries.  Once they drop support, it will help their sales of new Macs which will help their bottom line!  It will also kill the Hackintosh.  The real question is when.  I suspect that Apple will drop support for x86_64 binaries in three or four releases out.  Before then Apple will start letting developers know that support for X86_64 binaries will be dropped, so they will not be too shocked.
